rust/kernel/net/tcp/cong.rs

```diff
@@ -0,0 +1,544 @@
+use crate::init::PinInit;
+use crate::prelude::{pr_err, vtable};
+use crate::str::CStr;
+use crate::types::Opaque;
+use crate::ThisModule;
+use crate::{build_assert, field_size, try_pin_init};
+use core::convert::TryFrom;
+use core::marker::PhantomData;
+use core::pin::Pin;
+use macros::{pin_data, pinned_drop};
+
+use super::{InetConnectionSock, TcpSock};
+
+/// Congestion control algorithm (CCA).
+///
+/// A CCA is implemented as a set of callbacks that are invoked whenever
+/// specific events occur in a connection. Each socket has its own instance of
+/// some CCA. Every instance of a CCA has its own private data that is stored in
+/// the socket and is mutated by the callbacks.
+///
+/// Callbacks that operate on the same instance are guaranteed to run
+/// sequentially, and each callback has exclusive mutable access to the private
+/// data of the instance it operates on.
+#[vtable]
+pub trait Algorithm {
+    /// Private data. Each socket has its own instance.
+    type Data: Default + Send + Sync;
+
+    /// Name of the algorithm.
+    const NAME: &'static CStr;
+
+    /// Called when entering CWR, Recovery, or Loss states from Open or Disorder
+    /// states. Returns the new slow start threshold.
+    fn ssthresh(sk: &mut Sock<Self>) -> u32;
+
+    /// Called towards the end of processing an ACK if a cwnd increase is
+    /// possible. Performs a new cwnd calculation and sets it on the socket.
+    // Note: In fact, one of `cong_avoid` and `cond_control` is required.
+    // (see `tcp_validate_congestion_control`)
+    fn cong_avoid(sk: &mut Sock<Self>, ack: u32, acked: u32);
+
+    /// Called before the sender's congestion state is changed.
+    fn set_state(sk: &mut Sock<Self>, new_state: State) {}
+
+    /// Called when removing ACKed packets from the retransmission queue. Can be
+    /// used for packet ACK accounting.
+    fn pkts_acked(sk: &mut Sock<Self>, sample: &AckSample) {}
+
+    /// Called to undo a recent cwnd reduction that was found to has been
+    /// unnecessary. Returns the new value of cwnd.
+    fn undo_cwnd(sk: &mut Sock<Self>) -> u32;
+
+    /// Initializes the private data.
+    ///
+    /// When this function is called, [`sk.inet_csk_ca()`] will contain a value
+    /// returned by `Self::Data::default()`.
+    ///
+    /// Only implement this function when you need to perform additional setup
+    /// tasks.
+    ///
+    /// [`sk.inet_csk_ca()`]: Sock::inet_csk_ca
+    fn init(sk: &mut Sock<Self>) {}
+
+    /// Cleans up the private data.
+    ///
+    /// After this function returns, [`sk.inet_csk_ca()`] will be dropped.
+    ///
+    /// Only implement this function when you need to perform additional cleanup
+    /// tasks.
+    ///
+    /// [`sk.inet_csk_ca()`]: Sock::inet_csk_ca
+    fn release(sk: &mut Sock<Self>) {}
+}
+
+pub mod reno {
+    //! TCP Reno congestion control.
+    //!
+    //! Algorithms may choose to invoke these callbacks instead of providing
+    //! their own implementation.
```

Comment:
Could you add a quick blurb about why someone might use these?

Also undo_cwnd needs a doc comment.